viewing_party/party.py

```diff
@@ -1,23 +1,143 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if title == None or genre == None or rating == None:
+        return None
+    return {"title": title, "genre": genre, "rating": rating}
 
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, movie_title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == movie_title:
+            add_to_watched(user_data, movie)
+            user_data["watchlist"].remove(movie)
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    rating_sum = 0
+    counter = 0
+    try:
+        for movie in user_data["watched"]:
+            rating_sum += movie["rating"]
+            counter += 1
+        return rating_sum / counter
+    except ZeroDivisionError:
+        return 0
 
+def get_most_watched_genre(user_data):
+    # get and store list of all genre values
+    genre_total_list = []
+    for movie in user_data["watched"]:
+        genre_total_list.append(movie["genre"])
+    # create dictionary of most frequent genres
+    genre_frequencies = {}
+    for genre in genre_total_list:
+        try:
+            genre_frequencies[genre] += 1
+        except KeyError: 
+            genre_frequencies[genre] = 1
+    # identify genre with highest frequency count
+    most_popular_genre = None
+    max_frequency = 0
+    for genre, frequency in genre_frequencies.items():
+        if frequency > max_frequency:
+            most_popular_genre = genre
+            max_frequency = frequency
+    return most_popular_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    # start with list of all movies user has watched
+    movie_list = list(user_data["watched"])
+    unique_movies = list(movie_list)
+    # compare against list of friend's watched movies
+    friends = list(user_data["friends"])
+    for friend in friends:
+        for friend_movie in friend["watched"]:
+            if friend_movie in movie_list:
+                try:
+                    unique_movies.remove(friend_movie)
+                #handle a movie that's already been removed
+                except ValueError: 
+                    continue
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    #create list of movies friends have watched
+    friend_movies = []
+    for friend in user_data["friends"]:

```

Comment:
There's some shared logic between this and the last function that could be extracted into a helper function. It also has the same sort of strategy as the last function, so I have the same concern about shallow copying the list.